docs/proposals/20221205-cpu-schedule-latency.md

```diff
@@ -0,0 +1,264 @@
+
+*cilium/ebpf* is a pure Go library that provides utilities for loading, compiling, and debugging eBPF programs. It has minimal external dependencies and is intended to be used in long running processes.
+
+### Goals
+
+- Define structures, BPF Maps, and working mechanism for eBPF CPU schedule latency collector. 
+- Explain workflow of Koordinator and CPU schedule latency collector. 
+
+### Non-Goals/Future Work
+
+- Design external metrics collector framework to run metric collection without Koordlet invasion.
+- Define CRD or API for the collector to upload CPU schedule latency. 
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster administrator, I want to make use of Koordinator's Interference Detection feature to gain better cluster performance. 
+
+### Design
+
+#### Computation
+
+The CPU Schedule Latency is defined as below. Tracepoint `tp/sched/sched_wakeup` and `tp/sched/sched_wakeup_new` indicate one task is enqueued into `runqueue`.  And tracepoint `tp/sched/sched_switch` means one task finishes context switch. 
+
+![cpu-schedule-latency](../images/cpu-schedule-latency.svg)
+
+Notice that there might be many tasks and multiple times of context switch per task in one cgroup. So to compute CPU Schedule Latency on cgroup dimension, we sum all tasks' latency and record the number of latency reported by collector. The final `cgroup's CPU Schedule Latency` is defined as `sum(latency)/num(latency)`. 
+
+For further analysis and more information, the `sum` or `count` is also necessary to report when we want to check the accumulated latency of a cgroup. And based on that, the percentile values are also recorded using an algorithm just like Prometheus `HISTOGRAMS`: users can config buckets' boundaries for CPU schedule latency with `le(less equal)`, and the collector use cumulative counter to record each observation into corresponding bucket. 
+
+To compute one percentile of all observations, use belowing function: 
+
+```go
+bucketStart + (bucketEnd-bucketStart)*float64(rank/count)
+```
+
+For example, to compute P99 of 10000 observations, we need to know which bucket the 9900th observation is placed in. With cumulative buckets, suppose the 7th bucket contains 9850 records and the 8th bucket contains 9950, we can say that 99% records are under 8th buckets' upper bound( more precisely between 7th upper and 8th upper). Then we know the 9900th record is rank 50 in the 8th bucket, which has 100 counts in it. If the gap of each bucket is 0.2 seconds, then the `bucketStart` equals to `7*0.2`, and `bucketEnd-bucketStart` equals to `0.2`. Then we get the estimated P99 as: 
+
```

Comment:
Is it possible to configure the sampling ratio to reduce the collection overhead?

Reply:
Sure and it is worth doing. I updated some illustration about this in the newest commit.